rfc/merkle-structure.md

```diff
@@ -0,0 +1,320 @@
+# Abstract
+
+Inter Planetary Linked Data (IPLD) is a fundamental building block of the IPFS. IPLD Links (a.k.a CIDs) are quintessential, yet I'd argue a major design flaw (on par with null pointers ?)
+
+## Problems
+
+### Data Locality
+
+In most programming languages we build data structures out of primitive data types and more complex structures by nesting them into one another through [reference]s. In languages with managed memory you do not worry about whether you can dereference sub-structure reference in itself is implies access to it. Languages where you do need to worry about dereferencing tend to suffer from [null pointer] problems.
+
+Data structures in IPLD suffer from a similar set of problems as languages with
+[null pointer]s. In case of IPLD it is a locality problem. Nested sub-structures can either be **local**, inlined into a parent structure, or **external** linked using an IPLD link. Local references can always be dereferenced, but external references may require read from a disk, or another machine. In worse cases even searching the network to find machine that has it. Possibility of errors is inconvenient, but what's worse is the impact on the data model, imposing constraints of data exchange (block size) which are almost never universal.
```

Comment:
...do CIDs impose limits on block size? I think this specific implementation does, but that's not a hard constraint, right?

Reply:
CIDs do not impose limits but in practice limits are implied. New version of the PR goes into more detail on this.